sdk/communication/AzureCommunicationChat/Source/ChatClient.swift

```diff
@@ -330,4 +333,154 @@ public class ChatClient {
+                completionHandler(.failure(AzureError.client("Failed to start push notifications", error)))
+            }
+        }
+    }
+
+    /// Stop push notifications.
+    /// - Parameter completionHandler: Success indicates push notifications have been stopped.
+    public func stopPushNotifications(
+        completionHandler: @escaping (Result<HTTPResponse?, AzureError>) -> Void
+    ) {
+        guard pushNotificationsStarted else {
+            completionHandler(.failure(AzureError.client("Push notifications are not enabled.")))
+            return
+        }
+
+        // Recreate the registrarClient if it doesn't exist
+        if registrarClient == nil {
+            // Internal options do not use the CommunicationSignalingErrorHandler
+            let internalOptions = AzureCommunicationChatClientOptionsInternal(
+                apiVersion: AzureCommunicationChatClientOptionsInternal.ApiVersion(options.apiVersion),
+                logger: options.logger,
+                telemetryOptions: options.telemetryOptions,
+                transportOptions: options.transportOptions,
+                dispatchQueue: options.dispatchQueue
+            )
+
+            let communicationCredential = TokenCredentialAdapter(credential)
+            let authPolicy = BearerTokenCredentialPolicy(credential: communicationCredential, scopes: [])
+
+            // Initialize the RegistrarClient
+            guard let tempRegistrarClient = try? RegistrarClient(
+                endpoint: RegistrarSettings.endpoint,
+                credential: credential,
+                registrationId: registrationId,
+                authPolicy: authPolicy,
+                withOptions: internalOptions
+            ) else {
+                completionHandler(.failure(
+                    AzureError
+                        .client("Failed to stop push notifications. Fail to find or create registrarClient.")
+                ))
+                return
+            }
+
+            registrarClient = tempRegistrarClient
+        }
+
+        registrarClient!.deleteRegistration { result in
+            switch result {
+            case let .success(response):
+                self.pushNotificationsStarted = false
+                completionHandler(.success(response))
+            case let .failure(error):
+                self.options.logger.error("Failed to stop push notifications with error: \(error.localizedDescription)")
+                completionHandler(.failure(AzureError.client("Failed to stop push notifications", error)))
+            }
+        }
+    }
+
+    /// Handle the data payload for an incoming push notification.
+    /// - Parameters:
+    ///   - notification: The APNS push notification.
+    ///   - completionHandler: Receives the PushNotificationEvent from the push notification payload.
+    public func handlePush(
+        notification: [AnyHashable: Any],
+        completionHandler: PushNotificationEventHandler
```

Comment:
Something is definitely off about this method. Why does `startPushNotifications` not simply accept the event handler there? The fact that this is public is particularly weird since converting from data to an event type should be an internal operation, not something exposed to customers.

Reply:
This design is determined by the Push Notification Flow -

Let's imagine user A and user B are in the same chat thread. If user A sends a message, this message will go though a series of services and will finally reach user B's IOS device. Then user B's IOS device will forward this notification payload to Chat SDK. That's why we will need such a public function to accept the incoming notification payload. After receiving the payload, we will do some data conversion and data handler here. To handle the data, we will need an event handler passed by customer. In current codes, the customer directly passes "the eventHandler" together with "the notification they received" into this handlePush() API, in order to treat the data in a customized way.

Reply:
We can revisit this, since this is intended for preview. @mpodwysocki has some ideas on how this could be streamlined using method swizzling.